ceno_host/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+// See `make_stdin` and `consume` for the main entry points, and how this would look
+// for the host and guest respectively for the user of our library.
+// Everything else in here would be hidden.
+
+use itertools::izip;
+use rkyv::{
+    Portable, Serialize,
+    api::high::{HighSerializer, HighValidator},
+    bytecheck::CheckBytes,
+    rancor::Error,
+    ser::allocator::ArenaHandle,
+    to_bytes,
+    util::AlignedVec,
+};
+
+#[derive(Default)]
+pub struct CenoStdin {
+    pub items: Vec<AlignedVec>,
+}
+
+pub struct SerialisedCenoStdin(pub AlignedVec);
+
+impl CenoStdin {
+    pub fn write_slice(&mut self, bytes: AlignedVec) {
+        self.items.push(bytes);
+    }
+
+    pub fn write(
+        &mut self,
+        item: &impl for<'a> Serialize<HighSerializer<AlignedVec, ArenaHandle<'a>, Error>>,
+    ) -> Result<(), Error> {
+        let bytes = to_bytes::<Error>(item)?;
+        self.write_slice(bytes);
+        Ok(())
+    }
+
+    pub fn finalise(&self) -> SerialisedCenoStdin {
```

Comment:
Sorry, this is a bit of a mess.  I think it might be comparatively easier to understand the memory format by looking at the readers first, and then you can try to decipher this function.  (Or you can wait for the cleanup up version. :)

The main complexity here comes from me having mixed all the padding logic throughout.  What we are actually doing ain't very complicated conceptually.